### PR TITLE
Discard bad connections before Begin and Prepare, and improve cleanup in Close.

### LINKED ISSUE #58: Bad connections discovered during Tx.Commit or Tx.Rollback get reused in future transactions.
Although this is probably a bug in `database/sql` (https://github.com/golang/go/issues/11264), it has been around for a long time and may not get fixed there since other drivers are unaffected or have already worked around the issue. Even if it is fixed in the next release of Go, we may want to fix it here to work correctly with Go 1.5 and earlier.

### PR DESCRIPTION
**Pull request by ChrisHines**
Note that I also added connection tracking in the test code. That invaded the other tests, for which I apologize. Please suggest a better approach if you have one.

Fixes #58.

### CODE REVIEW

File: tx.go [hunk 1]
```diff
@@ -49,8 +51,10 @@ func (c *Conn) endTx(commit bool) error {
 	}
 	ret := api.SQLEndTran(api.SQL_HANDLE_DBC, api.SQLHANDLE(c.h), howToEnd)
 	if IsError(ret) {
-		return NewError("SQLEndTran", c.h)
+		err := c.newError("SQLEndTran", c.h)
```
Comment (alexbrainman) [inline tx.go]: Why create new variable? Just

return c.newError(...)

File: conn.go [hunk 2]
```diff
@@ -36,12 +37,31 @@ func (d *Driver) Open(dsn string) (driver.Conn, error) {
 	return &Conn{h: h}, nil
 }
 
-func (c *Conn) Close() error {
+func (c *Conn) Close() (err error) {
+	if c.tx != nil {
```
Comment (alexbrainman) [inline conn.go]: Why are you changing (*Conn).Close? Wouldn't ODBC rollback transaction if you release handle without committing it?

**Comment by alexbrainman** [comment]
New test fails in on both my linux:

=== RUN   TestMSSQLRecoverFromCommitBadConn                                           
--- FAIL: TestMSSQLRecoverFromCommitBadConn (0.01s)                                   
        mssql_test.go:1598: unexpected error return: should=driver: bad connection, is=SQLEndTran: {01000} [FreeTDS][SQL Server]Unexpected EOF from the server              
        mssql_test.go:1605: unexpected error return: should=driver: bad connection, is=already in a transaction                                                             
        mssql_test.go:129: error closing DB: driver: bad connection                   
FAIL                                                                                  

and some windows environment:

--- FAIL: TestMSSQLRecoverFromCommitBadConn (0.03s)
        mssql_test.go:1605: unexpected error return: should=driver: bad connection, is=SQLDriverConnect: {08001} [Microsoft][SQL Server Native Client 10.0]Client unable to establish connection because an error was encountered during handshakes before login. Common causes include client attempting to connect to an unsupported version of SQL Server, server too busy to accept new connections or a resource limitation (memory or maximum allowed connections) on the server.
                {08001} [Microsoft][SQL Server Native Client 10.0]TCP Provider: An existing connection was forcibly closed by the remote host.

                {08001} [Microsoft][SQL Server Native Client 10.0]Client unableto establish connection
                {08001} [Microsoft][SQL Server Native Client 10.0]Client unableto establish connection due to prelogin failure
        mssql_test.go:132: unexpected ConnCount: should <= 0, is=1
=== RUN   TestFoxPro
--- PASS: TestFoxPro (0.06s)
FAIL
exit status 1
FAIL    github.com/alexbrainman/odbc    1.359s

Alex

File: tx.go (hunk 1 above)
Comment (ChrisHines) [inline tx.go]: No reason, that is left over from a debugging session. I will fix it.

File: conn.go (hunk 2 above)
Comment (ChrisHines) [inline conn.go]: I found that if the transaction is not closed `api.SQLFreeHandle` returns the error "SQLFreeHandle: {HY010} [Microsoft][ODBC Driver Manager] Function sequence error".

**Comment by ChrisHines** [comment]
I will try to test on Linux and work on tightening up the test code.

File: conn.go (hunk 2 above)
Comment (alexbrainman) [inline conn.go]: Fair enough.

**Comment by ChrisHines** [comment]
@alexbrainman I simplified by testing directly against the driver interfaces instead of indirectly through the database/sql API. This version passes on for me on Windows 7 (against SQL Server 10 & 11) and CentOS 6 (against SQL Server 11).

**Comment by alexbrainman** [comment]
Thanks again.

Alex

**Pull request merged**